File: app/src/main/java/com/example/a8_puzzle/TilesAdapter.java
package com.example.a8_puzzle;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.TextView;

import java.util.List;

public class TilesAdapter extends BaseAdapter {
    
    private static final String TAG = "TilesAdapter";
    private final List<Tile> tiles;
    private final Context context;
    
    public TilesAdapter(List<Tile> tiles, Context context) {
        this.tiles = tiles;
        this.context = context;
    }
    
    @Override
    public int getCount() {
        if (tiles == null) return 0;
        return tiles.size();
    }
    
    @Override
    public Object getItem(int position) {
        return tiles.get(position);
    }
    
    @Override
    public long getItemId(int position) {
        return position;
    }
    
    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        // get tile
        final Tile tile = tiles.get(position);
        
        // instantiate a new cell view using the tile layout file
        if (convertView == null) {
            final LayoutInflater layoutInflater = LayoutInflater.from(context);
            convertView = layoutInflater.inflate(R.layout.tile_layout, null);
        }
        // set the text to the number of the tile
        final TextView tileNumber = convertView.findViewById(R.id.tile_layout_number);
        tileNumber.setText(String.valueOf(tile.getNumber()));
        
        // make the blank tile invisible
        if (tile.getNumber() == 9) {
            tileNumber.setVisibility(View.INVISIBLE);
            tileNumber.setClickable(false);
        }
        return convertView;
    }
}

File: app/src/main/java/com/example/cool8puzzle/TilesAdapter.kt
package com.example.cool8puzzle

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.TextView

class TilesAdapter(
    private val tiles: List<Int>,
    private val context: Context
) : BaseAdapter() {

    companion object {
        private const val TAG = "TilesAdapter"
    }

    override fun getCount(): Int {
        return tiles.size
    }

    override fun getItem(position: Int): Any {
        return tiles[position]
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup?): View? {
        // get tile
        var temp = convertView
        val tile = tiles[position]

        // instantiate a new cell view using the tile layout file
        // if convertView is null, assign it; otherwise return it
        if (temp == null) {
            val layoutInflater = LayoutInflater.from(context)
            temp = layoutInflater.inflate(R.layout.tile_layout, parent, false)
        }

        // set the text to the number of the tile
        val tileNumber = temp!!.findViewById<TextView>(R.id.tile_layout_number)
        tileNumber.text = tile.toString()

        // make the blank tile invisible
        if (tile == 9 || tile == 0) {
            tileNumber.visibility = View.INVISIBLE
            tileNumber.isClickable = false
        }
        return temp
    }

}